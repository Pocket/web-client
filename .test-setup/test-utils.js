// test-utils.js
// https://testing-library.com/docs/react-testing-library/api/
import { render, queries } from '@testing-library/react'
import * as customQueries from './custom-queries'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import { rootReducer } from '../src/store'
import { ViewportProvider } from '@pocket/web-ui'

/**
 * Custom basic render
 * @param {*} ui react element
 * @param {*} options passed in options from test-library
 * @returns
 */
const customRender = (ui, options) => {
  return render(ui, { queries: { ...queries, ...customQueries }, ...options })
}

/**
 * Render wrapped in provider for when you need that.  Use this sparingly since
 * tests really should be running in isolation
 * @param {*} ui react element
 * @param {*} options passed in options from test-library
 * @returns
 */
const wrappedRender = (ui, renderOptions = {}) => {
  const {
    initialState,
    store = createStore(rootReducer, initialState),
    router,
    ...options
  } = renderOptions

  const wrapper = ({ children }) => {
    return (
      <ViewportProvider>
        <Provider store={store}>{children}</Provider>
      </ViewportProvider>
    )
  }

  return render(ui, { wrapper, queries: { ...queries, ...customQueries }, ...options })
}

/**
 * Adding custom queries to screen element so we can follow best practices and get some
 * auto completions
 */
const boundQueries = Object.entries(customQueries).reduce((queries, [queryName, queryFn]) => {
  queries[queryName] = queryFn.bind(null, document.body)
  return queries
}, {})
const customScreen = { ...screen, ...boundQueries }

// re-export everything
export * from '@testing-library/react'

// export our enhanced elements
export { customScreen as screen, customRender as render, wrappedRender }

//Google ReCaptcha
// https://gist.github.com/Bartuz/baa4cb3cff45286a519acc5ad0313bd7
const nop = () => {}

const reaptchaCallbackName = (type) => {
  switch (type) {
    case 'success':
      return 'callback'
      break
    case 'error':
      return 'error-callback'
    case 'expire':
      return 'expired-callback'
    default:
      throw `Unknown callback type ${type}. Supported types: success, error, expire`
  }
}

export const mockGrecaptcha = ({
  forceCallbackType = 'success',
  reset = nop,
  execute = nop,
  token = 'g-token'
} = {}) => {
  window.grecaptcha = {
    ready: (callback) => callback(),
    render: (_html, reaptchaConfig) => {
      const callback = reaptchaConfig[reaptchaCallbackName(forceCallbackType)]
      callback(token)
    },
    reset: reset,
    execute: execute
  }
}

export const mockModal = () => {
  // Mock modal
  const setAppElementStub = jest.fn()
  const ReactModalMock = ({ children }) => <div>{children}</div>
  ReactModalMock.setAppElement = setAppElementStub
  jest.mock('react-modal', () => ReactModalMock)

  let portalRoot = document.getElementById('portal')
  if (!portalRoot) {
    const portalRoot = document.createElement('div')
    portalRoot.setAttribute('id', 'root')
    document.body.appendChild(portalRoot)
  }
  return { setAppElementStub }
}
