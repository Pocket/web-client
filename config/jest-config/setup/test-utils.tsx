import { RenderOptions, render } from '@testing-library/react'
import { ViewportProvider } from '../../../clients/web/src/components/viewport-provider/viewport-provider'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import { rootReducer, initializeStore } from '../../../clients/web/src/store'

import { i18n } from './test-i18n'
import { I18nextProvider } from 'react-i18next'

import { type JSX } from 'react'
import type { PropsWithChildren, ReactElement } from 'react'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
type RootState = ReturnType<typeof rootReducer>
type AppStore = ReturnType<typeof initializeStore>
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  initialState?: Partial<RootState>
  store?: AppStore
}

export function customRender(
  ui: ReactElement,
  {
    initialState,
    store = createStore(rootReducer, initialState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<NonNullable<unknown>>): JSX.Element {
    return (
      <I18nextProvider i18n={i18n}>
        <ViewportProvider>
          <Provider store={store}>{children}</Provider>
        </ViewportProvider>
      </I18nextProvider>
    )
  }
  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

//Google ReCaptcha
// https://gist.github.com/Bartuz/baa4cb3cff45286a519acc5ad0313bd7
declare global {
  interface Window {
    grecaptcha: any
  }
}

const nop = () => {}

const reaptchaCallbackName = (type: string) => {
  switch (type) {
    case 'success':
      return 'callback'
      break
    case 'error':
      return 'error-callback'
    case 'expire':
      return 'expired-callback'
    default:
      throw `Unknown callback type ${type}. Supported types: success, error, expire`
  }
}

export const mockGrecaptcha = ({
  forceCallbackType = 'success',
  reset = nop,
  execute = nop,
  token = 'g-token'
} = {}) => {
  window.grecaptcha = {
    ready: (callback: () => void) => callback(),
    render: (_html: any, reaptchaConfig: { [x: string]: any }) => {
      const callback = reaptchaConfig[reaptchaCallbackName(forceCallbackType)]
      callback(token)
    },
    reset: reset,
    execute: execute
  }
}

export const mockModal = () => {
  // Mock modal
  const setAppElementStub = jest.fn()
  const ReactModalMock = ({ children }: PropsWithChildren) => <div>{children}</div>
  ReactModalMock.setAppElement = setAppElementStub
  jest.mock('react-modal', () => ReactModalMock)

  const portalRoot = document.getElementById('portal')
  if (!portalRoot) {
    const portalRoot = document.createElement('div')
    portalRoot.setAttribute('id', 'root')
    document.body.appendChild(portalRoot)
  }
  return { setAppElementStub }
}

// Mock the IntersectionObserver, see https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API
export class IntersectionObserver {
  root = null
  rootMargin = ''
  thresholds = []

  disconnect() {
    return null
  }

  observe() {
    return null
  }

  takeRecords() {
    return []
  }

  unobserve() {
    return null
  }
}
window.IntersectionObserver = IntersectionObserver
global.IntersectionObserver = IntersectionObserver
