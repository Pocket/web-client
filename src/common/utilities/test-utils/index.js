/**
 * Returns an object with a data attribute key with id value that can be used to
 * apply as a prop on an element. Use the object spread operator to apply to an
 * element, like <div {...testIdAttribute('foo')}>, which will result in the div
 * having an attribute data-test-id="foo". The object is only returned if the
 * environment is not prod. The data attribute can be used to identify elements
 * in our tests.
 * @param  {String} id  id for element
 * @return {Object}     Object where key is the data attribute name, and the
 *                      value is the id.
 */
export function testIdAttribute(id) {
  if (process.env.NODE_ENV === 'test' || process.env.SHOW_DEV === 'included') {
    return { 'data-test-id': id }
  }
  return undefined
}

/**
 * Given a test id, returns a CSS attribute selector which will find the element
 * with that id. Useful with enzyme's `find()` function, or cypress' `cy.get()`
 *
 * @param   {String}  id         Test id of the element, it's data-test-id attribute value.
 * @param   {String}  qualifier  Query qualifier: ^ = starts with, * = matches any, $ = ends with.
 *                               See document.querySelector API.
 *
 * @return  {String}             The formatted test id attribute for use as a DOM selector.
 *                               e.g. [data-test-id*="article-card-"] which will match any
 *                               test id beggining with "article-card-".
 */
export function testIdSelector(id, qualifier = '') {
  if (qualifier && !['^', '*', '$'].includes(qualifier)) {
    throw new Error(
      `used an invalid qualifier for testIdSelector: "${qualifier}" (from ${id})`
    )
  }
  return `[data-test-id${qualifier}="${id}"]`
}

/**
 * Mock default event object that has some standard event things in it to prevent
 * handler functions from failing when simulating events
 * @type {Object}
 */
export const mockEvent = {
  target: {},
  preventDefault() {},
  stopPropagation() {},
}

/**
 * Create a mock response object based on what the fetch API would pass to its
 * Promise resolution.
 *
 * @param   {Object}   data    Any data that would be returned as parsed JSON
 * @param   {Boolean}  isFail  Set to true to indicate it's a failed request, which
 *                             will provide appropriate response values
 *
 * @return  {Object}          Mock response object to pass to Promise resolution
 */
export function createMockFetchResponse(data, isFail = false) {
  const response = {
    headers: {
      get(headerName) {
        return this[headerName]
      },
    },
  }

  if (data) {
    response.json = () => {
      return data
    }
  }

  if (isFail) {
    response.ok = false
    response.statusText = 'fetch error status text'
  } else {
    response.ok = true
  }

  return response
}
