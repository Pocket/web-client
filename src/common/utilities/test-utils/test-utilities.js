/**
 * Mock default event object that has some standard event things in it to prevent
 * handler functions from failing when simulating events
 * @type {Object}
 */
export const mockEvent = {
  target: {},
  preventDefault() {},
  stopPropagation() {}
}

/**
 * Create a mock response object based on what the fetch API would pass to its
 * Promise resolution.
 *
 * @param   {Object}   data    Any data that would be returned as parsed JSON
 * @param   {Boolean}  isFail  Set to true to indicate it's a failed request, which
 *                             will provide appropriate response values
 *
 * @return  {Object}          Mock response object to pass to Promise resolution
 */
export function createMockFetchResponse(data, isFail = false) {
  const response = {
    headers: {
      get(headerName) {
        return this[headerName]
      }
    }
  }

  if (data) {
    response.json = () => {
      return data
    }
  }

  if (isFail) {
    response.ok = false
    response.statusText = 'fetch error status text'
  } else {
    response.ok = true
  }

  return response
}
