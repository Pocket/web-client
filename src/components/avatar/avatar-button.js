import React from 'react'
import PropTypes from 'prop-types'
import { css, cx } from '@emotion/css'

import Avatar from './avatar'

const buttonStyles = css`
  background: var(--color-actionPrimarySubdued);
  border-radius: 50%;
  margin: 0;
  padding: 0;
  color: var(--color-textPrimary);
  position: relative;

  span.default {
    transition: all 0.1s ease-out;
  }

  span.with-image {
    &::after {
      content: '';
      display: block;
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
      border-radius: 50%;
      transition: all 0.1s ease-out;
      opacity: 0;
    }
  }

  &:focus {
    outline: none;
    &::after {
      content: '';
      display: block;
      position: absolute;
      top: -4px;
      bottom: -4px;
      left: -4px;
      right: -4px;
      border-radius: 50%;
      border: 2px solid var(--color-actionPrimary);
    }
  }

  &:hover {
    background: none;

    & span.default {
      color: var(--color-textLinkHover);
    }

    & span.with-image {
      &::after {
        background-color: var(--color-actionPrimary);
        opacity: 0.38;
      }
    }
  }

  &:active {
    & span.default {
      color: var(--color-textLinkPressed);
    }

    & span.with-image {
      &::after {
        background-color: var(--color-actionPrimaryHover);
        opacity: 38%;
      }
    }
  }
`

/**
 * Component that wraps the avatar component in a <button> and handles the interaction
 * styles of that button.
 */
// eslint-disable-next-line react/display-name
export const AvatarButton = React.forwardRef(
  (
    { id = '', label = '', src = null, onClick = () => {}, className = null, size, ...rest },
    ref
  ) => {
    return (
      <button
        type="button"
        className={cx(buttonStyles, className && className)}
        title={label}
        onClick={onClick}
        {...rest}
        ref={ref}>
        <Avatar size={size} src={src} altText={label} data-cy={`avatar-button-avatar-${id}`} />
      </button>
    )
  }
)

AvatarButton.propTypes = {
  /**
   * Width/height of the avatar. Since the avatar will be square, this value is
   * both width and height.
   */
  size: PropTypes.string.isRequired,

  /**
   * Alt text to include as the button label of the Avatar for non-sighted users
   */
  label: PropTypes.string,

  /**
   * Provide an id if you have multiple instances per page and need unique
   * identifiers for tests.
   */
  id: PropTypes.string,

  /**
   * The src/url for the avatar image
   */
  src: PropTypes.string,

  /**
   * Callback called when the button is clicked.
   */
  onClick: PropTypes.func,

  /**
   * Classname to be applied to the outer node that can be used to override
   * or supplement styles.
   */
  className: PropTypes.string
}

