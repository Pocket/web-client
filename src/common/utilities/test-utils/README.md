# Test Utils

A set of utility functions that assist with unit tests of frontend components.

## API

### testIdAttribute()

Function that creates a `data-test-id` attribute to pass to a component that will be used within React components to flag a particular element for identification in unit tests. Accepts a single param `elementId` that should be a unique string value to identify that element. The html attribute will only be added if the `NODE_ENV` value is `test`, so your frontend app should be set up to set that variable when running unit tests using this util.

```js
import { testIdAttribute } from '@pocket/web-utilities/test-utils'

// component file
const MyComponent = ({ title }) => {
  return (
    <div>
      <span {...testIdAttribute('title')}>${title}</span>
      This is a sample component with a prop-based title above.
    </div>
  )
}
```

### testIdSelector()

Function that creates a DOM selection string that will correspond to the `data-test-id` attribute created with the `testIdAttribute` function. This function is used within the unit tests to find an element flagged for a test within a React component. Accepts a single param `elementId` that should match the id used with `testIdAttribute`.

```js
import { testIdSelector } from '@pocket/web-utilities/test-utils'

// unit test file
describe('MyComponent', () => {
  it('renders the title passed as a prop', () => {
    const result = shallow(<MyComponent title="this is the title" />)
    const title = result.find(testIdSelector('title'))

    assert.equal(title.text(), 'this is the title')
  })
})
```

### mockEvent

`mockEvent` is an object that mimics a DOM event object. This is useful when simulating events in unit tests so that you can provide an event instance that already has commonly used properties and methods like a real event, so that you don't have to mock those in every test. For example, your component code might have a line like `event.preventDefault()`; this mock event object contains a noop function for that method so that your code doesn't throw an error when you simulate a fake event.

```js
import { mockEvent } from '@pocket/web-utilities/test-utils'

// unit test file
describe('MyComponent', () => {
  it('calls the onClick callback when the button is clicked', () => {
    const spy = sinon.spy()
    const result = shallow(<MyComponent onClick={spy} />)
    const button = result.find(testIdSelector('cool-button'))

    button.simulate(mockEvent)

    assert(spy.calledOnce)
  })
})
```

### createMockFetchResponse()

Function that creates a mock `response` object that would be returned from a `fetch` call. Acceps the following arguments in order to specify the contents of the response:

- `data [Object|Array|String]`: JSON-compatible value that will be returned when calling `response.json()` from the `response` returned from `fetch`
- `isFail [Boolean]`: Set this to `true` to indicate that the request should fail, which will set `response.ok` to `false` and `response.statusText` to "fetch error status text". Defaults to `false`.

```js
import { createMockFetchResponse } from '@pocket/web-utilities/test-utils'

const mockResponse = createMockFetchResponse({ someDataKey: 'someValue' })

console.log(mockResponse.json())
// { someDataKey: 'someValue' }

mockResponse.headers['x-error-code'] = '100'
console.log(mockResponse.headers.get('x-error-code'))
// '100'
```
