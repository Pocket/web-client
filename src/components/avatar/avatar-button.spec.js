import { render, fireEvent } from 'test-utils'
import renderer from 'react-test-renderer'
import '@testing-library/jest-dom/extend-expect'

import { AvatarButton } from './avatar-button'

describe('Avatar', () => {
  const baseProps = {
    size: '123px',
    label: 'view your kitties'
  }

  it('Applies a title attribute to the button if props.label is provided', () => {
    const { getByTitle } = render(<AvatarButton {...baseProps} />)
    expect(getByTitle('view your kitties')).toBeInTheDocument()

    const tree = renderer
    .create(<AvatarButton {...baseProps} />)
    .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Calls the props.onClick callback when the button is clicked', () => {
    const handleClick = jest.fn()
    const { getByTitle } = render(<AvatarButton {...baseProps} onClick={handleClick} />)
    fireEvent.click(getByTitle('view your kitties'))
    expect(handleClick).toHaveBeenCalledTimes(1)
  })

  it('applies a css class name to the button element if props.className is provided', () => {
    const classProps = {
      ...baseProps,
      className: 'pickles'
    }
    const { getByTitle } = render(<AvatarButton {...classProps} />)
    expect(getByTitle('view your kitties')).toHaveClass('pickles')

    const tree = renderer
    .create(<AvatarButton {...classProps} />)
    .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
