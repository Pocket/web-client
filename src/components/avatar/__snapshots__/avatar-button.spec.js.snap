// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Applies a title attribute to the button if props.label is provided 1`] = `
<button
  className="b502wo9"
  onClick={[Function]}
  title="view your kitties"
  type="button"
>
  <span
    className="ajoca6s default"
    style={
      Object {
        "height": "123px",
        "width": "123px",
      }
    }
  >
    <span
      className="igxbmuu icon dwtruy8"
      data-cy="avatar-default-"
    >
      <svg
        aria-hidden="true"
        aria-labelledby=" "
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 11.125a5.067 5.067 0 01-5.06-5.063A5.067 5.067 0 0112 1c2.79 0 5.06 2.271 5.06 5.062A5.067 5.067 0 0112 11.125zM12 3a3.06 3.06 0 10.001 6.121A3.06 3.06 0 0012 3.001zM11.99 15.006c4.41 0 7.98 2.302 8.01 4.993H4v-.07c.12-2.741 3.63-4.923 7.99-4.923zm0-2c-5.44 0-9.85 3.071-9.99 6.883V20C2 21.1 2.9 22 4 22h16c1.09 0 1.98-.88 2-1.97-.04-3.883-4.5-7.025-10.01-7.025z"
        />
      </svg>
    </span>
  </span>
</button>
`;

exports[`Avatar applies a css class name to the button element if props.className is provided 1`] = `
<button
  className="b502wo9 pickles"
  onClick={[Function]}
  title="view your kitties"
  type="button"
>
  <span
    className="ajoca6s default"
    style={
      Object {
        "height": "123px",
        "width": "123px",
      }
    }
  >
    <span
      className="igxbmuu icon dwtruy8"
      data-cy="avatar-default-"
    >
      <svg
        aria-hidden="true"
        aria-labelledby=" "
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 11.125a5.067 5.067 0 01-5.06-5.063A5.067 5.067 0 0112 1c2.79 0 5.06 2.271 5.06 5.062A5.067 5.067 0 0112 11.125zM12 3a3.06 3.06 0 10.001 6.121A3.06 3.06 0 0012 3.001zM11.99 15.006c4.41 0 7.98 2.302 8.01 4.993H4v-.07c.12-2.741 3.63-4.923 7.99-4.923zm0-2c-5.44 0-9.85 3.071-9.99 6.883V20C2 21.1 2.9 22 4 22h16c1.09 0 1.98-.88 2-1.97-.04-3.883-4.5-7.025-10.01-7.025z"
        />
      </svg>
    </span>
  </span>
</button>
`;
